EIPS/eip-4762.md

```diff
@@ -39,47 +39,43 @@ We define access events as follows. When an access event takes place, the access
 
 #### Access events for account headers
 
-When a non-precompile address is the target of a `CALL`, `CALLCODE`, `DELEGATECALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, or is the target address of a contract creation whose initcode starts execution, process these access events:
+When:
 
-```
-(address, 0, VERSION_LEAF_KEY)
-(address, 0, CODE_SIZE_LEAF_KEY)
-```
+ 1. a non-precompile address is the target of a `CALL`, `CALLCODE`, `DELEGATECALL`, `STATICALL`, `AUTHCALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, 
+ 2. a non-precompile address is the target address of a contract creation whose initcode starts execution,
+ 3. any address is the target of the `BALANCE` opcode
+ 4. a _deployed_ contract calls `CODECOPY`
 
-If a call is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these two access events:
+process this access events:
 
 ```
-(caller_address, 0, BALANCE_LEAF_KEY)
-(callee_address, 0, BALANCE_LEAF_KEY)
+(address, 0, BASIC_DATA_LEAF_KEY)
 ```
 
-When a contract is created, process these access events:
+If a `*CALL` or `SELFDESTRUCT` is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these additional two access events:
 
 ```
-(contract_address, 0, VERSION_LEAF_KEY)
-(contract_address, 0, NONCE_LEAF_KEY)
-(contract_address, 0, BALANCE_LEAF_KEY)
```

Comment:
I'm wondering if converting balance from a `uint256` to a `uint128` might cause breaking changes in some contexts. To be backwards compatible, it also puts languages like Solidity in a position where they will have to treat balances as `uint256` (for e.g. the very commonly used `address.balance` returns a `uint256`.) even though the underlying data structure only contains a uint128. Ethereum's ecosystem also includes a variety of layer-2 solutions, sidechains, and off-chain protocols that may rely on the current uint256 account structure. Changing to uint128 could introduce incompatibilities with these systems.

Not saying we shouldn't do it, and I'm a fan of the simplification & more efficient packed data structure, but just pointing out these potential risks.

Reply:
well if balance can never exceed 16 bytes it doesn't matter since clients stack can still treat it as 32 bytes and only writing to and fro the state is 16 bytes

Reply:
according to etherscan/coinmarketcap, the total ether supply peaked above 120M, which is $120 \times 10^{18 + 6}$ wei, which corresponds to about 87 bits if my math is correct. To reach a situation where this is a problem, we need to multiply the issuance by $2^{42}$ and this would have to land at the same address.

Reply:
> I'm wondering if converting balance from a `uint256` to a `uint128` might cause breaking changes in some contexts. To be backwards compatible, it also puts languages like Solidity in a position where they will have to treat balances as `uint256` (for e.g. the very commonly used `address.balance` returns a `uint256`.) even though the underlying data structure only contains a uint128. Ethereum's ecosystem also includes a variety of layer-2 solutions, sidechains, and off-chain protocols that may rely on the current uint256 account structure. Changing to uint128 could introduce incompatibilities with these systems.
> 
> 
> 
> Not saying we shouldn't do it, and I'm a fan of the simplification & more efficient packed data structure, but just pointing out these potential risks.

Reply:
I think we should include a mechanism to handle balances larger than 128bit, even if it is not used. The overflow balance could be stored in an 'additional balance leaf' which will be accessed if the 128bit one is maxed out. This makes the gas cost for reading balance dynamic though, which could be another source of contract bugs, though we wouldn't be hitting this in practice on eth mainnet.